web/packages/teleport/src/Roles/RoleEditor/standardmodel.ts

```diff
@@ -53,6 +54,24 @@ export type MetadataModel = {
   revision?: string;
 };
 
+export type AccessSpec = KubernetesAccessSpec | ServerAccessSpec;
+
+type AccessSpecBase<T extends AccessSpecKind> = {
+  /**
+   * Determines kind of resource that is accessed using this spec. Intended to
+   * be mostly consistent with UnifiedResources.kind, but that has no real
+   * meaning on the server side; we needed some discriminator, so we picked
+   * this one.
+   */
+  kind: T;
+};
+
+export type KubernetesAccessSpec = AccessSpecBase<'kube_cluster'> & {};
+
+export type ServerAccessSpec = AccessSpecBase<'node'> & {};
+
+export type AccessSpecKind = AccessSpec['kind'];
```

Comment:
@gzdunek Access spec base is, well, a base. It may actually be a bit of an overkill, but I thought it may be a good thing to say in one place that "this is a type that holds a kind" and give it a common documentation comment. Then, we have access specs specific to Kubernetes, Server access, apps, databases, etc., that correspond to various sections of the UI. The `AccessSpec` doesn't refer to itself, it's just a union of these types. I noticed that you added this comment on an old commit; perhaps it will be more clear if you look at the last one.

This way or another, I'll explain it more clearly in the docs.

Reply:
I just looked at the last commit and I guess I still don't fully understand it:
1. `AccessSpec` is a union of `KubernetesAccessSpec` and `ServerAccessSpec`.
2. `KubernetesAccessSpec` and `ServerAccessSpec` extend `AccessSpecBase`.
3. `AccessSpecBase` is a generic type that extends `AccessSpecKind`.
4. `AccessSpecKind` depends on `AccessSpec['kind']` - so we go back to point 1.

So don't we go into a loop? It feels to me that `AccessSpecBase` should look like this: 
```
type AccessSpecBase<T> = {
  /**
   * Determines kind of resource that is accessed using this spec. Intended to
   * be mostly consistent with UnifiedResources.kind, but that has no real
   * meaning on the server side; we needed some discriminator, so we picked
   * this one.
   */
  kind: T;
};
```
`AccessSpecBase` is a building block for `KubernetesAccessSpec` and `ServerAccessSpec` so it shouldn't refer to them. Let me know if I'm missing something.

Reply:
Oh, now I see what you're talking about! Well, yes, in this case, we are going in a loop, but that works, and it's a pattern used in at least one more place in our codebase. I guess I could specify `AccessSpecKind` explicitly, though.

Reply:
But what does `<T extends AccessSpecKind>` give us?
I can do this 
<img width="591" alt="image" src="https://github.com/user-attachments/assets/ef28ca83-843e-4c3a-a597-f67588a10c3e">

And TS doesn't complain because `T extends AccessSpecKind` doesn't provide any restriction for the type at this point.